_posts/2019-10-23-kafka-authentication-using-oauth-2.0.md

```diff
@@ -0,0 +1,538 @@
+
+
+<!--more-->
+
+## Authentication in Kafka
+
+When using Apache Kafka you have several options to implement authentication. 
+You can use TLS with client certificates to establish a secure channel and authenticate a client at the same time.
+Another option is to separately configure TLS for connection encryption and integrity, but use a specific SASL protocol for authentication.
+
+There are several SASL based protocols you can use.
+One option is to use SASL PLAINTEXT to authenticate with username and password.
+Another option is to use SASL SCRAM which is also username and password based, but rather than storing clear text passwords it stores salted hashes.
+These two implementations both store usernames and credentials in ZooKeeper.
+
+A more advanced, enterprise-ready solution is to use SASL GSSAPI mechanism which provides support for Kerberos. 
+That typically uses a central LDAP server for managing users, and permissions.
+The problem is that it is quite complex to setup, and people often find it hard to configure.
+
+Since Kafka version 2.0.0 there is an extensible OAuth 2.0 compatible token-based mechanism available, called SASL OAUTHBEARER.
+
+We have developed extensions that provide integration with OAuth 2.0 compliant authorization servers. 
+That means that in principle you can use any OAuth 2.0 compliant authorization server to expose centrally managed users for authentication with Kafka.
+
+
+## Why OAuth 2.0?
+
+Originally developed for scenarios where web users need to give limited access to their data inside one web site to another web site, OAuth 2.0 provides at least two main benefits.
+One is that user accounts and credentials for many services are managed centrally, organization-wide, by a dedicated identity service.
+Another important benefit is that the user only ever passes credentials like username and password to that dedicated server.
+The result of authenticating is a time-limited token which the user passes around to other services, and which serves to identify user, and to establish which permissions they have.
+ 
+
+## Hands-on
+
+Let's see how this works. As our OAuth 2.0 compliant authorization server we will use a basic setup of Keycloak with TLS support.
+We'll then configure a Kafka cluster with OAuth 2.0 support, and an example Kafka producer, and Kafka consumer client applications, that will use OAuth 2.0 to authenticate. 
+
+We'll assume that you have a Kubernetes cluster available in order to explain how this works for Strimzi, but the OAuth 2.0 support can be used with non-Strimzi Kafka installations too, by using [Strimzi Kafka OAuth libraries](https://github.com/strimzi/strimzi-kafka-oauth).
+ 
+If you are using `minikube` on your local machine, make sure to start `minikube tunnel` in another terminal in order for LoadBalancer services to get their externals IPs assigned:
+
+    sudo minikube tunnel 
```

Comment:
The Netlify preview doesn't render the blog post because  you have the date in the future. But does Jekyll render this properly?  I think I never tried this before. GitHub seems to have some issues rendering the text, but it might work fine with Jekyll. You should check it anyway to make sure abotu it before we publish it.

Reply:
I found that simply changing header in .md file to have date in the past makes it show:
...
date: 2019-10-22
...
It may take a minute for browser to pick it up.

Reply:
Netlify generates the preview - but I cannot change it easily in your branch. Anyway, the important thing is if the formating works well.